### ISSUE TITLE
Arrow history

### ISSUE THREAD

**Issue by steveoh** [chrisant996/clink#275]
Hi I updated from an old cmder to the latest and I am having a really hard time getting history to work. It worked out of the box I thought in the old cmder but not any longer.

I've installed the latest clink but still no history. There is a history file in the directory but it seems partial. Is there anything obvious from this information that can help us get history back?

![image](https://user-images.githubusercontent.com/325813/164345333-55f13500-969f-4f05-8b87-222da1ff00c6.png)

**Comment by chrisant996** [comment]
Could you elaborate?

What does "doesn't work" mean?  What _does_ happen instead?

Are you expecting all of the old Clink v0.4.9 history to still exist, and the old history is missing?

Or do you mean that new commands are not saved in history?

What is the output from `clink set`?

Please attach the `clink.log` file from `c:\tools\Cmder\config\clink.log`.

**Comment by steveoh** [comment]
yes, I can elaborate - sorry. Nothing happens when i use the up arrows after i run a command. If i use the searches, my command is not found. I cannot access any history. 

The clink install should have all the defaults as far as I know since I haven't changed any settings.

<details>
<summary>clink set</summary>

```
clink set
autosuggest.async                 True
autosuggest.enable                True
autosuggest.original_case         True
autosuggest.strategy              match_prev_cmd history completion
clink.autostart
clink.colorize_input              True
clink.default_bindings            windows
clink.gui_popups                  False
clink.logo                        full
clink.paste_crlf                  crlf
clink.path
clink.promptfilter                True
cmd.admin_title_prefix
cmd.auto_answer                   off
cmd.ctrld_exits                   False
cmd.get_errorlevel                True
color.arg                         bold
color.arginfo                     sgr 38;5;172
color.argmatcher                  sgr 1;38;5;40
color.cmd                         sgr 1;38;5;231
color.cmdredir                    sgr 38;5;172
color.cmdsep                      sgr 38;5;214
color.comment_row                 sgr 38;5;87;48;5;18
color.description                 sgr 38;5;39
color.doskey                      sgr 1;38;5;75
color.executable                  sgr 1;38;5;33
color.filtered                    sgr 38;5;231
color.flag                        sgr 38;5;117
color.git.star                    bright green
color.hidden                      sgr 38;5;160
color.horizscroll                 sgr 38;5;16;48;5;30
color.input                       sgr 38;5;222
color.interact                    bold
color.message                     default
color.modmark
color.popup
color.popup_desc
color.prompt
color.readonly                    sgr 38;5;28
color.selected_completion         sgr 38;5;16;48;5;254
color.selection                   sgr 38;5;16;48;5;179
color.suggestion                  sgr 38;5;239
color.unexpected                  default
color.unrecognized                sgr 38;5;203
debug.log_terminal                False
doskey.enhanced                   True
exec.aliases                      True
exec.commands                     True
exec.cwd                          True
exec.dirs                         True
exec.enable                       True
exec.files                        False
exec.path                         True
exec.space_prefix                 True
files.hidden                      True
files.system                      False
history.dont_add_to_history_cmds  exit history
history.dupe_mode                 erase_prev
history.expand_mode               not_dquoted
history.ignore_space              False
history.max_lines                 10000
history.save                      True
history.shared                    True
history.sticky_search             False
lua.break_on_error                False
lua.break_on_traceback            False
lua.debug                         False
lua.path
lua.reload_scripts                False
lua.strict                        True
lua.traceback_on_error            False
match.expand_envvars              True
match.fit_columns                 True
match.ignore_accent               True
match.ignore_case                 relaxed
match.max_fitted_matches          0
match.max_rows                    0
match.preview_rows                5
match.sort_dirs                   with
match.substring                   True
match.translate_slashes           system
match.wild                        True
prompt.async                      True
prompt.transient                  off
readline.hide_stderr              False
terminal.adjust_cursor_style      True
terminal.differentiate_keys       False
terminal.east_asian_ambiguous     auto
terminal.emulation                auto
terminal.mouse_input              auto
terminal.mouse_modifier
terminal.raw_esc                  False
terminal.use_altgr_substitute     False
```
</details>

<details>
<summary>clink.log</summary>

```
1364 start_logger              154 ---- 2022/02/09 12:36:38.653 -------------------------------------------------
1364 start_logger              160 Host process is 'cmd.exe' (pid 4964)
1364 start_logger              164 DLL path is 'C:\tools\Cmder\vendor\clink'
1364 start_logger              178 Windows version 10.0.17763 (x64)
1364 start_logger              181 Clink version 1.2.46.69fc92 (x64)
1364 hook_setter::hook_setter  156 >>> Started hook transaction.
1364 hook_setter::attach_iat   313 Attempting to hook SetEnvironmentVariableW in IAT for module 00007FF77D130000.
1364 pe_info::iterate_imports  160 Found import in 'api-ms-win-core-processenvironment-l1-1-0.dll'
1364 find_iat                  121 Found import at 00007FF77D161830 (value is 00007FFFF47AEAF0).
1364 hook_setter::attach_iat   313 Attempting to hook WriteConsoleW in IAT for module 00007FF77D130000.
1364 pe_info::iterate_imports  160 Found import in 'api-ms-win-core-console-l1-1-0.dll'
1364 find_iat                  121 Found import at 00007FF77D1614C8 (value is 00007FFFF4795820).
1364 hook_setter::attach_iat   313 Attempting to hook GetEnvironmentVariableW in IAT for module 00007FF77D130000.
1364 pe_info::iterate_imports  160 Found import in 'api-ms-win-core-processenvironment-l1-1-0.dll'
1364 find_iat                  121 Found import at 00007FF77D161850 (value is 00007FFFF475EFA0).
1364 hook_setter::commit       204 <<< Hook transaction committed.
1364 hook_setter::hook_setter  156 >>> Started hook transaction.
1364 hook_setter::detach_iat   356 Attempting to unhook 00007FFFCEE0ADF0 from GetEnvironmentVariableW in IAT for module 00007FF77D130000.
1364 pe_info::iterate_imports  160 Found import in 'api-ms-win-core-processenvironment-l1-1-0.dll'
1364 find_iat                  121 Found import at 00007FF77D161850 (value is 00007FFFCEE0ADF0).
1364 hook_setter::commit       204 <<< Hook transaction committed.
1364 hook_setter::hook_setter  156 >>> Started hook transaction.
1364 hook_setter::attach_iat   313 Attempting to hook ReadConsoleW in IAT for module 00007FF77D130000.
1364 pe_info::iterate_imports  160 Found import in 'api-ms-win-core-console-l1-1-0.dll'
1364 find_iat                  121 Found import at 00007FF77D1614B0 (value is 00007FFFF48314D0).
1364 hook_setter::commit       204 <<< Hook transaction committed.
1364 reset_handle              145 resetting mismatched stdout handle
1364 win_screen_buffer::begin  227 Using native terminal support; found 'ConEmuHk64.dll'.
1364 history_db::initialise   1268 master bank ctag: |CTAG_1644431101_499062_6620_0
1364 history_db::compact      1492 History:  7 active, 2 deleted
1364 reset_handle              145 resetting mismatched stdout handle
1364 history_db::compact      1492 History:  8 active, 2 deleted
1364 reset_handle              145 resetting mismatched stdout handle
1364 history_db::compact      1492 History:  9 active, 2 deleted
1364 reset_handle              145 resetting mismatched stdout handle
1364 history_db::compact      1492 History:  10 active, 2 deleted
1364 reset_handle              145 resetting mismatched stdout handle
1364 history_db::compact      1492 History:  11 active, 2 deleted
1364 reset_handle              145 resetting mismatched stdout handle
1364 history_db::compact      1492 History:  12 active, 2 deleted
1364 history_db::compact      1492 History:  13 active, 2 deleted
1364 history_db::compact      1492 History:  14 active, 2 deleted
1364 reset_handle              145 resetting mismatched stdout handle
1364 history_db::compact      1492 History:  15 active, 2 deleted
1364 reset_handle              145 resetting mismatched stdout handle
1364 history_db::compact      1492 History:  16 active, 2 deleted
1364 reset_handle              145 resetting mismatched stdout handle
1364 history_db::compact      1492 History:  17 active, 2 deleted
1364 reset_handle              145 resetting mismatched stdout handle
1364 history_db::compact      1492 History:  18 active, 2 deleted
1364 history_db::compact      1492 History:  19 active, 2 deleted
1364 reset_handle              145 resetting mismatched stdout handle
1364 history_db::compact      1492 History:  20 active, 2 deleted
1364 reset_handle              145 resetting mismatched stdout handle
1364 history_db::compact      1492 History:  21 active, 2 deleted
1364 reset_handle              145 resetting mismatched stdout handle
1364 history_db::compact      1492 History:  22 active, 2 deleted
1364 history_db::compact      1492 History:  23 active, 2 deleted
1364 reset_handle              145 resetting mismatched stdout handle
1364 history_db::compact      1492 History:  24 active, 2 deleted
1364 reset_handle              145 resetting mismatched stdout handle
1364 history_db::compact      1492 History:  25 active, 2 deleted
1364 history_db::compact      1492 History:  26 active, 2 deleted
1364 history_db::compact      1492 History:  27 active, 2 deleted
1364 reset_handle              145 resetting mismatched stdout handle
1364 history_db::compact      1492 History:  28 active, 2 deleted
1364 history_db::compact      1492 History:  29 active, 2 deleted
1364 reset_handle              145 resetting mismatched stdout handle
1364 history_db::compact      1492 History:  29 active, 3 deleted
1364 history_db::compact      1492 History:  29 active, 4 deleted
1364 history_db::compact      1492 History:  30 active, 4 deleted
1364 history_db::compact      1492 History:  30 active, 5 deleted
1364 history_db::compact      1492 History:  30 active, 6 deleted
1364 history_db::compact      1492 History:  31 active, 6 deleted
```
</details>

**Comment by chrisant996** [comment]
The log shows that commands are being added successfully to history.  You can tell because the "active" count changes/grows many times in the log.

What happens when you run `history` or `clink history`?  Based on the log, I expect it will list the commands in the history.

The next thing to do is run `clink echo` and then press the <kbd>Up</kbd> key, and see if something at the terminal level is intercepting the key.  E.g. ConEmu/Cmder and various other terminals allow key bindings that override Clink.  (Press <kbd>Ctrl</kbd>-<kbd>C</kbd> to exit from `clink echo`, when finished.)

Also, try pressing <kbd>Alt</kbd>-<kbd>Shift</kbd>-<kbd>/</kbd> to invoke the `clink-what-is` command, and then press <kbd>Up</kbd> to double check what command the key is bound to.

**Comment by steveoh** [comment]
clink echo looks normal

```
λ clink echo
Type a key to see its key sequence string.
Press Ctrl+C when finished.
"\e[A"
"\e[B"
"\C-C"
```

history crashes

<details>
<summary>history and clink history both crash</summary>

```
λ history

!!! CLINK'S CRASHED!
!!! v1.3.15.6e6e45 (x64)
!!!
!!! Writing core dump
!!! C:\tools\Cmder\config\clink.dmp
!!! ...ok
!!!
!!! Clink: 0x00007FFC1D1E0000
!!! Host: 0x00007FF6BF570000
!!! Guru: 0xc0000005 addr=0x0000000000000000 (1)
!!!
!!! Backtrace:
!!! 0x00007FFC1D1F1C25
!!! 0x00007FFC1D1F1E13
!!! 0x00007FFC1D1E7DC5
!!! 0x00007FFC1D1E8CFB
!!! 0x00007FFC1D1E4FAC
!!! 0x00007FFC1D1E48EE
!!! 0x00007FF6BF57100A
!!! 0x00007FFC44CB7974
!!! 0x00007FFC44DEA2F1

Press Enter to exit...
```
</details>

I'm remote desktop'd in from a mac so i can't get the what-is command to trigger. as it seems alt isn't mapping properly to my option or fn key.

**Comment by chrisant996** [comment]
Oh, I missed an important detail:  you shared some old clink.log from February.  The date in the log file is `2022/02/09 12:36:38.653` which means it's not the right log file (it's more than 2 months old).  Also, it's using `Clink version 1.2.46.69fc92 (x64)`, but you indicated you are using Clink version 1.3.15.6e6e45.

It sounds like there are some configuration and/or permission problems that are external to Clink, and are interfering with Clink and/or Cmder.  I'm not sure how to help walk through specific troubleshooting steps.

Are you sure that the _only_ thing you did was extract the latest Cmder-mini.zip contents over the top of the previous Cmder installation?

> history crashes
> history and clink history both crash

That's interesting, and is likely to be informative.
Can you put the `C:\tools\Cmder\config\clink.dmp` core dump file into a `.zip` file and upload it?

> I'm remote desktop'd in from a mac so i can't get the what-is command to trigger. as it seems alt isn't mapping properly to my option or fn key. Is there another way to invoke that command?

You could bind it to a different key, by creating an `.inputrc` file.
See [Init File](https://chrisant996.github.io/clink/clink.html#init-file).

This sample `.inputrc` file would make <kbd>Ctrl</kbd>-<kbd>/</kbd> invoke the `clink-what-is` command.
```inputrc
"\e[27;5;191~": clink-what-is    # Ctrl-/
```

**Comment by steveoh** [comment]
I installed cmder back in february in a fresh new windows server and when history wasn't available i installed clink based on some github issue recommendations. It seems that clink may be installed with cmder or historically that was the case? 

`clink-what-is` seems to be correct `Up : previous-history -- Move 'back' through the history list, fetching the previous command`

[clink.zip](https://github.com/chrisant996/clink/files/8535634/clink.zip)

**Comment by chrisant996** [comment]
The user account does not have permission to the profile directory and/or to the clink_history file.

This is a permissions problem.  I'll fix the crash when there are no permissions, but you will have to fix the permissions problem.

**Comment by steveoh** [comment]
yep, fooled with the permissions on that folder and i now have history. Thanks!

**Comment by chrisant996** [comment]
The crash is fixed for v1.3.16, and also now `history` and `clink history` report an error message if the history file(s) can't be opened.  That will make it easier for users to realize when there are permission problems or other system configuration errors.
